cmd/influx/config/config.go

```diff
@@ -95,34 +161,161 @@ func (svc LocalConfigsSVC) WriteConfigs(pp Configs) error {
+// ListConfigs decodes configs from io readers
+func (s baseRW) ListConfigs() (Configs, error) {
+	cfgs := make(Configs)
+	_, err := toml.DecodeReader(s.r, &cfgs)
+	for n, cfg := range cfgs {
+		cfg.Name = n
+		cfgs[n] = cfg
+	}
+	return cfgs, err
+}
+
+// CreateConfig create new config.
+func (svc localConfigsSVC) CreateConfig(cfg Config) (Config, error) {
+	if cfg.Name == "" {
+		return Config{}, &influxdb.Error{
+			Code: influxdb.EInvalid,
+			Msg:  "config name is empty",
+		}
+	}
+	cfgs, err := svc.ListConfigs()
+	if err != nil {
+		return Config{}, err
+	}
+	if _, ok := cfgs[cfg.Name]; ok {
+		return Config{}, &influxdb.Error{
+			Code: influxdb.EConflict,
+			Msg:  fmt.Sprintf("config %q already exists", cfg.Name),
+		}
+	}
+	cfgs[cfg.Name] = cfg
+	if err := cfgs.Switch(cfg.Name); err != nil {
+		return Config{}, err
+	}
+
+	return cfgs[cfg.Name], svc.writeConfigs(cfgs)
+}
+
+// DeleteConfig will delete a config.
+func (svc localConfigsSVC) DeleteConfig(name string) (Config, error) {
+	cfgs, err := svc.ListConfigs()
+	if err != nil {
+		return Config{}, err
+	}
+
+	p, ok := cfgs[name]
+	if !ok {
+		return Config{}, &influxdb.Error{
+			Code: influxdb.ENotFound,
+			Msg:  fmt.Sprintf("config %q is not found", name),
+		}
+	}
+	delete(cfgs, name)
+
+	return p, svc.writeConfigs(cfgs)
+}
+
+// SwitchConfig will active the config by name, if name is "-", active the previous one.
+func (svc localConfigsSVC) SwitchConfig(name string) (Config, error) {
+	var up Config
+	if name == "-" {
+		p0, err := svc.parsePreviousActive()
+		if err != nil {
+			return Config{}, err
+		}
+		up.Name = p0.Name
+	} else {
+		up.Name = name
+	}
+	up.Active = true
+	return svc.UpdateConfig(up)
+}
+
+// UpdateConfig will update the config.
+func (svc localConfigsSVC) UpdateConfig(up Config) (Config, error) {
+	cfgs, err := svc.ListConfigs()
+	if err != nil {
+		return Config{}, err
+	}
+	p0, ok := cfgs[up.Name]
+	if !ok {
+		return Config{}, &influxdb.Error{
+			Code: influxdb.ENotFound,
+			Msg:  fmt.Sprintf("config %q is not found", up.Name),
+		}
+	}
+	if up.Token != "" {
+		p0.Token = up.Token
+	}
+	if up.Host != "" {
+		p0.Host = up.Host
+	}
+	if up.Org != "" {
+		p0.Org = up.Org
+	}
+
+	cfgs[up.Name] = p0
+	if up.Active {
+		if err := cfgs.Switch(up.Name); err != nil {
+			return Config{}, err
+		}
+	}
+
+	return cfgs[up.Name], svc.writeConfigs(cfgs)
+}
+
+// writeConfigs to the path.
+func (s ioStore) writeConfigs(cfgs Configs) error {
+	if err := os.MkdirAll(s.Dir, os.ModePerm); err != nil {
+		return err
+	}
+	var b1 bytes.Buffer
+	if err := (baseRW{w: &b1}).writeConfigs(cfgs); err != nil {
+		return err
+	}
+	return ioutil.WriteFile(s.Path, b1.Bytes(), 0600)
+}
+
+// parsePreviousActive return the previous active config from the reader
+func (s baseRW) parsePreviousActive() (Config, error) {
+	return s.parseActiveConfig(false)
 }
 
 // ParseActiveConfig returns the active config from the reader.
 func ParseActiveConfig(r io.Reader) (Config, error) {
```

Comment:
curious why is this exported? Wont' the service encapsulate this?

Reply:
Looks like it's been used by `getConfigFromDefaultPath`, looks like it will ignore some IO errors, so not something i feel comfortable to change this time. No enough testing coverage for this.

Reply:
I'm not following the concern here 🤔

Reply:
the concern is that has nothing to do the `switch back` pr, behavior change should be a separated issue, with separated tests